src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHostUserInterfaceProgress.cs

```diff
@@ -166,6 +217,10 @@ class ConsoleHostUserInterface : System.Management.Automation.Host.PSHostUserInt
 
         private ProgressPane _progPane = null;
         private PendingProgress _pendingProgress = null;
+        // The timer set up 'progPaneUpdateFlag' every 'UpdateTimerThreshold' milliseconds to update 'ProgressPane'
+        private Timer _progPaneUpdateTimer;
```

Comment:
An issue with showing progress bar on the same thread with a timer is that it's possible there is still activity records not shown and `progPaneUpdateFlag` is true, but there is no more `Write-Progress` calls. In such case, some progress records will never be shown. Is there a way to guarantee all pending progress records to be shown before `Reset` is called? Or maybe do another `Show` in `Reset` when `_pendingProgress` is not empty?